zdm/grid.py

```diff
@@ -0,0 +1,672 @@
+            kz2=r/pzc[iz2]
+            kz1=1.-kz2
+            MCz=self.zvals[iz2]*kz2
+            pDM=pzDM[iz2,:] # just use the value of lowest bin
+        
+        
+        # NOW DO dm
+        #pDM=pzDM[k,:]
+        pDMc=np.cumsum(pDM)
+        pDMc /= pDMc[-1]
+        r=np.random.rand(1)[0]
+        iDM2=np.where(pDMc>r)[0][0]
+        if iDM2 > 0:
+            iDM1=iDM2-1
+            dDM=r-pDMc[iDM1]
+            kDM2=dDM/(pDMc[iDM2] - pDMc[iDM1])
+            kDM1=1.-kDM2
+            MCDM=self.dmvals[iDM1]*kDM1 + self.dmvals[iDM2]*kDM2
+            if iz2>0:
+                Eth=self.thresholds[j,iz1,iDM1]*kz1*kDM1 \
+                    + self.thresholds[j,iz1,iDM2]*kz1*kDM2 \
+                    + self.thresholds[j,iz2,iDM1]*kz2*kDM1 \
+                    + self.thresholds[j,iz2,iDM2]*kz2*kDM2
+            else: 
+                Eth=self.thresholds[j,iz2,iDM1]*kDM1 \
+                    + self.thresholds[j,iz2,iDM2]*kDM2
+                Eth *= kz2**2 #assume threshold goes as Eth~z^2 in the near Universe
+        else:
+            # interpolate linearly from 0 to the minimum value
+            kDM2=r/pDMc[iDM2]
+            MCDM=self.dmvals[iDM2]*kDM2
+            if iz2>0: # ignore effect of lowest DM bin on threshold
+                Eth=self.thresholds[j,iz1,iDM2]*kz1 \
+                    + self.thresholds[j,iz2,iDM2]*kz2
+            else: 
+                Eth=self.thresholds[j,iz2,iDM2]*kDM2
+                Eth *= kz2**2 #assume threshold goes as Eth~z^2 in the near Universe
+        
+        # now account for beamshape
+        Eth /= MCb
+        
+        # NOW GET snr
+        #Eth=self.thresholds[j,k,l]/MCb
+        Es=np.logspace(np.log10(Eth),np.log10(self.Emax),100)
+        PEs=self.vector_cum_lf(Es,self.Emin,self.Emax,self.gamma)
+        PEs /= PEs[0] # normalises: this is now cumulative distribution from 1 to 0
+        r=np.random.rand(1)[0]
+        iE1=np.where(PEs>r)[0][-1] #returns list starting at 0 and going upwards
+        iE2 = iE1+1
+        # iE1 should never be the highest energy, since it will always have a probability of 0
+        kE1=(r-PEs[iE2])/(PEs[iE1]-PEs[iE2])
+        kE2=1.-kE1
+        MCE=10**(np.log10(Es[iE1])*kE1 + np.log10(Es[iE2])*kE2)
+        MCs=MCE/Eth
+        
+        FRBparams=[MCz,MCDM,MCb,j,MCs]
+        return FRBparams,pwb
+        
+
+    def update(self, vparams:dict):
+        """[summary]
+        
+        Hierarchy:
+        Each indent corresponds to one 'level'.
+        This is used in the program control below
+        to dictate how far each tree should proceed
+        in calculation.
+        Direct variable inputs are always listed first
+        We see that sfr evolution and dm smearing
+        lie just before the pdv step
+        Hence, we deal with these first, and
+        calc rates as a final step regardless
+        of what else has changed.
+        
+        calc_rates:
```

Comment:
We should update this help to reflect the new structure with H0. I am OK making this a to-do for now - perhaps leave a note in the code to this effect however?

Reply:
made a few edits